### ISSUE TITLE
[🐛 bug report] Crash when rendering twice in a row with AD (reference counting)

### ISSUE THREAD

**Issue by merlinND** [mitsuba-renderer/mitsuba3#113]
## Summary

When enabling AD gradients and rendering the `classroom` scene twice in a row with the path tracer, I encounter the following crash:

## System configuration

*For bug report, please enter information regarding your system configuration<- [remove this]*

- Platform: Ubuntun 20.04
- Compiler: Clang 12
- Python version: 3.8.10
- Mitsuba 3 version: e4b1979f
- Compiled variants:
    * `scalar_rgb`
    * `cuda_ad_rgb`

## Steps to reproduce

```python
from os.path import join, realpath

import mitsuba as mi
import drjit as dr

def render_scene(scene_path, render_params, use_ad):
    if False:
        import gc
        # This doesn't influence the crash
        gc.collect()
        gc.collect()
        dr.eval()
        dr.sync_thread()
        dr.flush_malloc_cache()
        dr.malloc_clear_statistics()

    scene = mi.load_file(scene_path, **render_params)
    if use_ad:
        params = mi.traverse(scene)
        params.keep(r'.*reflectance\.(value|data)')
        assert len(params) >= 1
        for k, v in params:
            dr.enable_grad(v)
    else:
        params = None

    image = mi.render(scene, params=params, spp=render_params['spp'])

    if use_ad:
        loss = dr.mean(dr.sqr(image - 0.5))
        # Doesn't crash if call to `backward` is omitted
        dr.backward(loss)


if __name__ == '__main__':
    mi.set_variant('cuda_ad_rgb')

    scenes_dir = realpath(join(__file__, '..', '..', 'scenes'))
    scene_path = join(scenes_dir, 'classroom', 'scene.xml')

    render_params = {
        'integrator': 'path', 'resx': 256, 'resy': 256, 'spp': 4,
    }
    use_ad = True

    # Does not crash in the kitchen scene or when not using AD
    # scene_path = join(scenes_dir, 'kitchen', 'scene.xml')
    # use_ad = False

    render_scene(scene_path, render_params, use_ad)
    print(f'Render 1 OK')
    # Triggers regardless of AD here
    render_scene(scene_path, render_params, False)
    print(f'Render 2 OK')
```

## Description

The first call to `mi.render()` works fine, but the second crashes with:

```
Render 1 OK


Critical Dr.Jit compiler failure: cuda_check(): API error 0700 (CUDA_ERROR_ILLEGAL_ADDRESS): "an illegal memory access was encountered" in ../ext/drjit/ext/drjit-core/src/cuda_tex.cpp:70.
[1]    662766 abort      /usr/bin/python3 
```

It could be an issue with the reference counting or memory management logic.
